packages/contracts/src/dollar/BondingCurve.sol

```diff
@@ -0,0 +1,164 @@
+
+    event Withdraw(address indexed recipient, uint256 amount);
+
+    modifier onlyBondingMinter() {
+        require(
+            manager.hasRole(manager.BONDING_MINTER_ROLE(), msg.sender), "not minter"
+        );
+        _;
+    }
+
+    modifier onlyPauser() {
+        require(
+            manager.hasRole(manager.PAUSER_ROLE(), msg.sender), "not pauser"
+        );
+        _;
+    }
+
+    modifier onlyParamsSet() {
+        require(
+            connectorWeight != 0 && baseY != 0, "not initialised"
+        );
+        _;
+    }
+
+    constructor(
+        address _manager,
+        address _token
+    ) {
+        require(_token != address(0), "NFT address empty");
+        tokenAddr = _token;
+        token = IUbiquiStick(_token);
+
+        manager = UbiquityDollarManager(_manager);
+    }
+
+    function setCollateralToken(address _collateral) external onlyBondingMinter {
+        collateral = _collateral;
+    }
+
+    function setParams(
+        uint32 _connectorWeight, 
+        uint256 _baseY
+    ) external onlyBondingMinter {
+        require(_connectorWeight > 0 && _connectorWeight <= 1000000, "invalid values"); 
+        require(_baseY > 0, "must valid baseY");
+
+        connectorWeight = _connectorWeight;
+        baseY = _baseY;
+    }
+
+    /// @notice 
+    /// @dev 
+    /// @param _collateralDeposited Amount of collateral
+    /// @param _recipient An address to recieve the NFT
+    /// @return Tokens minted
+    function deposit(uint256 _collateralDeposited, address _recipient)
+        external
+        onlyParamsSet
+        returns (uint256)
+    {
+        uint256 tokensReturned;
+
+        if (tokenIds > 0) {
+            tokensReturned = _purchaseTargetAmount(
+                _collateralDeposited,
+                connectorWeight,
+                tokenIds,
+                poolBalance
+            );
+        } else {
+            tokensReturned = _purchaseTargetAmountFromZero(
+                _collateralDeposited,
+                connectorWeight,
+                ACCURACY,
+                baseY
+            );
+        }
+
+        IERC20(manager.dollarTokenAddress()).transferFrom(
+            msg.sender,
+            address(this),
+            _collateralDeposited
+        );
+
+        poolBalance += _collateralDeposited;
+        tokenIds += 1;
+        uint256 nftNum = 1;
+        bytes memory tokReturned = toBytes(tokensReturned);
+
+        token.mint(_recipient, tokenIds, nftNum, tokReturned);
+        emit Deposit(_recipient, _collateralDeposited);
+
+        return tokensReturned;
+    }
+
+    function toBytes(uint256 x) internal pure returns (bytes memory b) {
+        b = new bytes(32);
+        assembly { mstore(add(b, 32), x) }
+    }
+
+    function pause() public virtual onlyPauser {
+        _pause();
+    }
+
+    function unpause() public virtual onlyPauser {
+        _unpause();
+    }
+
+    function withdraw(address _to, uint256 _amount) external onlyBondingMinter {
```

Comment:
so it is a withdraw all collateral function. @pavlovcik don't you think we should restrict the withdraw to the DAO treasury (you can get the address from the manager)  ?

Reply:
I am just gonna add a function to set the treasure address.

Reply:
[Treasury](https://etherscan.io/address/fisc.ubq.eth) is a 4/4 multisig and is tedious to do operations with. The "admin" [wallet](https://etherscan.io/address/ubq.eth) is a MetaMask hot wallet which is far more convenient to do multi-transaction-operations with.

Could be nice to "withdraw to treasury" but I would consider the capability to withdraw to the "admin" wallet is more essential.

Reply:
@pavlovcik should i revert back to just a simple function that sets "admin" wallet?

Reply:
like I said no need to set treasury cos it is inside the Manager already you can do like this :
IUbiquityDollarManager(_manager).treasuryAddress()